# Syllabus (1-10 is fixed)

## 1. English title (free text field)

Metagenomics (2 ECTS)

## 2. Swedish title (free text field)

Metagenomik (2 ECTS)

## 3. Course credits (drop down menu: not possible to divide into modules here, so we must go for the total credits, comment from SLUBI)

2 ECTS

## 4. Subject (drop down menu: limited number of options but I guess Bioinformatics should be OK?, comment from SLUBI)

Bioinformatics

## 5. Course type (drop down menu: General course or Subject course, I suggest the latter, comment from SLUBI)

Subject course

## 6. Language (drop down menu: Swedish or English)
English

## 7. Prerequisites (free text field to describe knowledge/merits required for admission)

Basic knowledge of Unix (specifically, the command line interface).
The course is primarily for PhD but will be open for all interested PhD students/researchers


##  8. Objective (free text field)

[THIS WE NEED TO FILL]
The aims of this course are to:
    • familiarize the participants with advanced RNA-Seq data analysis methodologies and how biological knowledge can be gained from these by illustrating different analyses approaches.
    • allow the participants to acquire computational competences about latest RNA-Seq analytical approaches and related statistical methods
    • expose the participants to modern computing technologies, as the computer hands-on of the course will be performed using cloud computing resources
    • develop the participants’ awareness to critically reflect onto RNA-Seq analysis strengths and weaknesses
    • enable participants to perform the analysis of their own data (in stance of the exemplary dataset otherwise available)
[THIS WE NEED TO FILL]
The corresponding Learning Outcomes are:

	Module one (optional)
        ◦ develop basic skills to navigate and interact with the Unix Command Line Interface (CLI)
        ◦ develop basic skills to interact with the RStudio Integrated Development Environment (IDE)
        ◦ reproduce basic analysis in the R programming language

Module two
        ◦ summarise High Throughput Sequencing (HTS) technologies, past and present
        ◦ report advantages and limitations of HTS for expression profiling
        ◦ describe the HTS data pre-processing
        ◦ perform the HTS data pre-processing
        ◦ describe the pseudo-alignment principles
        ◦ compare (pseudo-)alignment methods
        ◦ perform the pseudo-alignment
        ◦ list statistical concepts of importance for RNA-Seq
        ◦ perform the biological QA
        ◦ describe the principle of Differential Expression (DE)
        ◦ enumerate the statistical concepts of DE
        ◦ list the assumptions associated with DE
        ◦ perform the differential expression analysis



## 9. Content (free text field)

[THIS WE NEED TO FILL]
High Throughput Sequencing (HTS) technologies and in particular Next Generation Sequencing (NGS) have revolutionized the way we conduct gene expression analysis. In comparison to micro-array-based methods, NGS has unleashed an almost unlimited power to perform gene expression analysis, for a similar price and with faster, more comprehensive, more efficient and more reproducible characteristics. Rapidly, the data generation rate has exceeded the analytical capabilities and data analysis has become the major bottleneck in gene expression studies. This advanced course, aimed at advanced students, has the objective to help addressing this problem by (i) helping participants critically assess the challenges faced in the HTS field, (ii) developing efficient communication skill with bioinformatician, and (iii) training students in analyzing HTS data.
The course has one main module, and two optional modules (two and one ECT each, respectively). The main module covers the analysis of RNA-Seq data, while the first optional module is an introduction to the Unix and R environments to give attendees the necessary admission requirements for the main module. The second optional module is a follow-up for attendees to apply the knowledge gained from the main module content on their own data. The course will provide two to four ECTs depending on the selected modules, as follows:

Module 1 (optional) 1 ECT
Module 2 (mandatory) 2 ECTs
Module 3 (optional) 1 ECT

Module 1 will be happening ahead of the main module and last two days. Literature reading and pre-course material for module 2 will be made available prior to the course. Module 2 and 3 will last 5 and 1 day(s), respectively.

For Module 3, the students who are interested will be able to analyse data from their own project by applying the methods learned in module 2 and with some guidance from the teachers (synchronous and asynchronous). Students will need to organise their access to the high-performance computing facility (probably their PI / supervisor will need to do so), with the help of SLUBI if needed. First, with asynchronous support from SLUBI, the students will perform the pre-processing and initial analysis of their data. Then, during a day, they will meet with the trainers to address issues, discuss the data interpretation and prepare a flash presentation about their project and results. These will then be given to the other participants and trainers in the form of an online mini-symposium.

The course will run for five to eight days; depending on whether optional modules are selected by the students - mixing lectures, interactive lectures, computer hands-on session, literature review, etc. The course will use various teaching environment, including meetings as a classroom, but also as smaller groups. In addition, asynchronous virtual environment will be used, such as virtual classrooms, flipped classrooms, etc. to offer the participants the possibility to review and deepen their understanding of the course material at their own pace and to discuss it among peers.

## 10. Examination (free text field)

[THIS WE NEED TO FILL]
The course will have a continuous assessment, conducted once a day at the end of the day, based on activities including, but not limited to, small group’s work with peer-review, literature review, journal club, etc. The continuous assessment and the final exam will provide five marks (six, seven or eight if either or both of the optional modules are also attended). All assessments must be passed in order to pass the course.

Requirement for examination
Attendance at all scheduled activities and approved computer exercises. All the material will be made available and students missing a day will be offered the possibility to go through it in self-study and be assessed.


## 11. Contact for application and further information (Name and E-mail)

[THIS WE NEED TO FILL]
Nicolas Delhomme (course leader): nicolas.delhomme@slu.se

## 12. Literature (free text field)

All the relevant material will be provided ahead of the course, typically a couple weeks ahead of the main module and will be made available after the course.

##  13. Additional information (free text field)

[THIS WE NEED TO FIX]
Teaching on this course will be provided by the SLU Bioinformatics Infrastructure (SLUBI), namely Nicolas Delhomme, Juliette Hayer and Domenico Simone. The course is organized by the SLU research school Organism Biology. There is no tuition fee. Participants are expected to bring their own laptop for the practical computer exercises.

Due to the current COVID-19 situation, the course will be conducted online only. However, if students wish so, they can organise themselves in groups and join using one of the teaching facilities available at SLU.

The course has a maximum number of participants is 20. PhD students will be prioritised, but admission will be on a first come, first serve basis, provided that the minimal requirements are met.

Responsible Department
Department of Aquatic Sciences and Assessment
